File: app/src/main/java/com/example/smsreaderpro/data/parser/EvcPlusParser.kt
package com.example.smsreaderpro.data.parser

import com.example.smsreaderpro.data.model.ParseResult
import com.example.smsreaderpro.data.model.TransactionType
import java.text.SimpleDateFormat
import java.util.Locale

class EvcPlusParser : TransactionParser {
    private val validSenders = listOf("EVCPlus", "192", "Hormuud", "waafi", "770", "780", "199", "141")

    override fun canParse(sender: String, body: String): Boolean {
        val s = sender.trim().lowercase(Locale.ROOT)
        val isSenderMatch = validSenders.any { s.contains(it.lowercase(Locale.ROOT)) || it.lowercase(Locale.ROOT).contains(s) }
        val hasEvcKeywords = body.contains("EVCPlus", ignoreCase = true) ||
                body.contains("waa laguu soo diray", ignoreCase = true) ||
                body.contains("waad u dirtay", ignoreCase = true) ||
                body.contains("Haraagaagu waa", ignoreCase = true) ||
                body.contains("Haraaga:", ignoreCase = true) ||
                body.contains("[+$") ||
                body.contains("[-$")
        return isSenderMatch || hasEvcKeywords
    }

    override fun parse(sender: String, body: String): ParseResult? {
        return try {
            val lower = body.lowercase(Locale.ROOT)
            var type = when {
                body.contains("[+$") || lower.contains("waa laguu soo diray") || lower.contains("waxaad ka heshay") -> TransactionType.INCOME
                lower.contains("ku shubatay") || lower.contains("kugu shubtay") || lower.contains("airtime") -> TransactionType.AIRTIME
                lower.contains("bixisay") || lower.contains("ku bixisay") || lower.contains("merchant") -> TransactionType.BILL_PAYMENT
                else -> TransactionType.EXPENSE
            }

            var amount = 0.0
            val bracketRegex = Regex("""\[([+-]?)\$([0-9.,]+)\]""")
            val bracketMatch = bracketRegex.find(body)
            if (bracketMatch != null) {
                amount = bracketMatch.groupValues[2].replace(",", "").toDoubleOrNull() ?: 0.0
                if (bracketMatch.groupValues[1] == "+") type = TransactionType.INCOME
                if (bracketMatch.groupValues[1] == "-") type = TransactionType.EXPENSE
            } else {
                val dollarRegex = Regex("""\$([0-9.,]+)""")
                val dollarMatch = dollarRegex.find(body)
                if (dollarMatch != null) {
                    amount = dollarMatch.groupValues[1].replace(",", "").toDoubleOrNull() ?: 0.0
                }
            }

            if (amount <= 0.0) return null

            var counterpartyName: String? = null
            var counterpartyPhone: String? = null

            if (type == TransactionType.INCOME) {
                val senderRegex = Regex("""(?:Waxaana soo diray|ka heshay)\s*([^(\n]+?)\s*(?:\(([^)]+)\))?""", RegexOption.IGNORE_CASE)
                val senderMatch = senderRegex.find(body)
                if (senderMatch != null) {
                    counterpartyName = senderMatch.groupValues[1].trim()
                    if (senderMatch.groupValues.size > 2) {
                        counterpartyPhone = senderMatch.groupValues[2].trim().ifEmpty { null }
                    }
                }
            } else {
                val receiverRegex = Regex("""(?:Waad u dirtay|u dirtay|u wareejisay)\s*([^(\n]+?)\s*(?:\(([^)]+)\))?""", RegexOption.IGNORE_CASE)
                val receiverMatch = receiverRegex.find(body)
                if (receiverMatch != null) {
                    counterpartyName = receiverMatch.groupValues[1].trim()
                    if (receiverMatch.groupValues.size > 2) {
                        counterpartyPhone = receiverMatch.groupValues[2].trim().ifEmpty { null }
                    }
                }
            }

            var balance: Double? = null
            val balanceRegex = Regex("""Haraaga(?:agu)?(?:\s+waa)?[:\s]*\$([0-9.,]+)""", RegexOption.IGNORE_CASE)
            val balanceMatch = balanceRegex.find(body)
            if (balanceMatch != null) {
                balance = balanceMatch.groupValues[1].replace(",", "").toDoubleOrNull()
            }

            var reference: String? = null
            val refRegex = Regex("""Ref[:\s]*([A-Za-z0-9]+)""", RegexOption.IGNORE_CASE)
            val refMatch = refRegex.find(body)
            if (refMatch != null) {
                reference = refMatch.groupValues[1].trim()
            }

            var timestamp = System.currentTimeMillis()
            val dateRegex = Regex("""Tar[:\s]*(\d{2}/\d{2}/\d{2,4}\s+\d{2}:\d{2}(?::\d{2})?)""", RegexOption.IGNORE_CASE)
            val dateMatch = dateRegex.find(body)
            if (dateMatch != null) {
                try {
                    val sdf = SimpleDateFormat("dd/MM/yy HH:mm:ss", Locale.ROOT)
                    val dateStr = dateMatch.groupValues[1].trim()
                    val parsed = sdf.parse(if (dateStr.count { it == ':' } == 1) "$dateStr:00" else dateStr)
                    if (parsed != null) timestamp = parsed.time
                } catch (_: Exception) {}
            }

            val counterparty = when {
                counterpartyName != null && counterpartyPhone != null -> "$counterpartyName ($counterpartyPhone)"
                counterpartyName != null -> counterpartyName
                else -> counterpartyPhone
            }

            var category = when (type) {
                TransactionType.INCOME -> "Dakhli"
                TransactionType.AIRTIME -> "Ku hadal Taleefan"
                TransactionType.BILL_PAYMENT -> "Biilasha"
                else -> "Other"
            }

            counterpartyName?.lowercase(Locale.ROOT)?.let { lowerName ->
                when {
                    lowerName.contains("bajaaj") || lowerName.contains("taxi") -> category = "Bajaaj"
                    lowerName.contains("cunto") || lowerName.contains("rest") || lowerName.contains("cafe") -> category = "Cunto"
                    lowerName.contains("dukaan") || lowerName.contains("supermarket") || lowerName.contains("mall") -> category = "Dukaan"
                    lowerName.contains("koronto") || lowerName.contains("electric") || lowerName.contains("beco") -> category = "Koronto"
                    lowerName.contains("biyo") || lowerName.contains("water") -> category = "Biyo"
                    lowerName.contains("kiro") || lowerName.contains("rent") -> category = "Kiro"
                    lowerName.contains("internet") || lowerName.contains("wifi") -> category = "Internet"
                    lowerName.contains("shaah") || lowerName.contains("tea") -> category = "Shaah"
                }
            }

            ParseResult(
                amount = amount,
                type = type,
                sender = if (type == TransactionType.INCOME) counterparty else null,
                receiver = if (type != TransactionType.INCOME) counterparty else null,
                balance = balance,
                reference = reference,
                provider = "EVCPlus",
                category = category,
                confidenceScore = 0.95f,
                timestamp = timestamp
            )
        } catch (_: Exception) {
            null
        }
    }
}

File: src/parser/EvcPlusParser.ts
import { TransactionParser } from './TransactionParser';
import { ParseResult, TransactionType } from '../types';

export class EvcPlusParser implements TransactionParser {
  private readonly validSenders = ['EVCPlus', '192', 'Hormuud', 'waafi', '770', '780', '199', '141'];

  canParse(sender: string, body: string): boolean {
    const s = sender.trim().toLowerCase();
    const isSenderMatch = this.validSenders.some(
      (valid) => s.includes(valid.toLowerCase()) || valid.toLowerCase().includes(s)
    );
    const hasEvcKeywords =
      body.includes('EVCPlus') ||
      body.includes('waa laguu soo diray') ||
      body.includes('waad u dirtay') ||
      body.includes('Haraagaagu waa') ||
      body.includes('Haraaga:') ||
      body.includes('[+$') ||
      body.includes('[-$');

    return isSenderMatch || hasEvcKeywords;
  }

  parse(sender: string, body: string): ParseResult | null {
    try {
      let type: TransactionType = 'EXPENSE';
      let amount = 0;
      let counterpartyName: string | undefined;
      let counterpartyPhone: string | undefined;
      let balance: number | undefined;
      let reference: string | undefined;
      let timestamp: number | undefined;

      // Check transaction type
      if (body.includes('[+$') || body.toLowerCase().includes('waa laguu soo diray') || body.toLowerCase().includes('waxaad ka heshay')) {
        type = 'INCOME';
      } else if (body.toLowerCase().includes('ku shubatay') || body.toLowerCase().includes('kugu shubtay') || body.toLowerCase().includes('airtime')) {
        type = 'AIRTIME';
      } else if (body.toLowerCase().includes('bixisay') || body.toLowerCase().includes('ku bixisay') || body.toLowerCase().includes('merchant')) {
        type = 'BILL_PAYMENT';
      } else {
        type = 'EXPENSE';
      }

      // Amount extraction
      const bracketAmountMatch = body.match(/\[([+-]?)\$([0-9.,]+)\]/);
      if (bracketAmountMatch) {
        amount = parseFloat(bracketAmountMatch[2].replace(/,/g, ''));
        if (bracketAmountMatch[1] === '+') type = 'INCOME';
        if (bracketAmountMatch[1] === '-') type = 'EXPENSE';
      } else {
        // Match general dollar patterns: $35.00 or 35.00 USD
        const dollarMatch = body.match(/\$([0-9.,]+)/) || body.match(/([0-9.,]+)\s*USD/i);
        if (dollarMatch) {
          amount = parseFloat(dollarMatch[1].replace(/,/g, ''));
        }
      }

      if (isNaN(amount) || amount <= 0) {
        return null;
      }

      // Counterparty extraction
      if (type === 'INCOME') {
        // "Waxaana soo diray FAARAX MAXAMED (615887766)"
        const senderMatch = body.match(/Waxaana soo diray\s*([^(\n]+?)\s*(?:\(([^)]+)\))?/i) ||
                            body.match(/ka heshay\s*([^(\n]+?)\s*(?:\(([^)]+)\))?/i);
        if (senderMatch) {
          counterpartyName = senderMatch[1]?.trim();
          counterpartyPhone = senderMatch[2]?.trim();
        }
      } else {
        // "Waad u dirtay DUKAANKA CUNTADA (612334455)"
        const receiverMatch = body.match(/Waad u dirtay\s*([^(\n]+?)\s*(?:\(([^)]+)\))?/i) ||
                              body.match(/u dirtay\s*([^(\n]+?)\s*(?:\(([^)]+)\))?/i) ||
                              body.match(/u wareejisay\s*([^(\n]+?)\s*(?:\(([^)]+)\))?/i);
        if (receiverMatch) {
          counterpartyName = receiverMatch[1]?.trim();
          counterpartyPhone = receiverMatch[2]?.trim();
        }
      }

      // Balance extraction: "Haraagaagu waa $185.00" or "Haraaga: $185.00"
      const balanceMatch = body.match(/Haraaga(?:agu)?(?:\s+waa)?[:\s]*\$([0-9.,]+)/i);
      if (balanceMatch) {
        const parsedBalance = parseFloat(balanceMatch[1].replace(/,/g, ''));
        if (!isNaN(parsedBalance)) {
          balance = parsedBalance;
        }
      }

      // Reference extraction: "Ref:EVC1234" or "Ref: 123456"
      const refMatch = body.match(/Ref[:\s]*([A-Za-z0-9]+)/i);
      if (refMatch) {
        reference = refMatch[1].trim();
      }

      // Date/Time extraction: "Tar: 16/09/26 19:10:00"
      const dateMatch = body.match(/Tar[:\s]*(\d{2}\/\d{2}\/\d{2,4}\s+\d{2}:\d{2}(?::\d{2})?)/i);
      if (dateMatch) {
        try {
          const parts = dateMatch[1].split(' ');
          const dateParts = parts[0].split('/');
          const timeParts = parts[1].split(':');
          let year = parseInt(dateParts[2], 10);
          if (year < 100) year += 2000;
          const month = parseInt(dateParts[1], 10) - 1;
          const day = parseInt(dateParts[0], 10);
          const hour = parseInt(timeParts[0], 10);
          const min = parseInt(timeParts[1], 10);
          const sec = timeParts[2] ? parseInt(timeParts[2], 10) : 0;
          timestamp = new Date(year, month, day, hour, min, sec).getTime();
        } catch {
          // ignore date parse fallback
        }
      }

      const counterparty = counterpartyName
        ? (counterpartyPhone ? `${counterpartyName} (${counterpartyPhone})` : counterpartyName)
        : counterpartyPhone;

      // Smart auto-category
      let category = type === 'INCOME' ? 'Dakhli' : 'Other';
      if (type === 'AIRTIME') category = 'Ku hadal Taleefan';
      if (type === 'BILL_PAYMENT') category = 'Biilasha';
      if (counterpartyName) {
        const lowerName = counterpartyName.toLowerCase();
        if (lowerName.includes('bajaaj') || lowerName.includes('taxi')) category = 'Bajaaj';
        else if (lowerName.includes('cunto') || lowerName.includes('rest') || lowerName.includes('cafe')) category = 'Cunto';
        else if (lowerName.includes('dukaan') || lowerName.includes('supermarket') || lowerName.includes('mall')) category = 'Dukaan';
        else if (lowerName.includes('koronto') || lowerName.includes('electric') || lowerName.includes('beco')) category = 'Koronto';
        else if (lowerName.includes('biyo') || lowerName.includes('water')) category = 'Biyo';
        else if (lowerName.includes('kiro') || lowerName.includes('rent')) category = 'Kiro';
        else if (lowerName.includes('internet') || lowerName.includes('wifi')) category = 'Internet';
        else if (lowerName.includes('shaah') || lowerName.includes('tea')) category = 'Shaah';
      }

      return {
        amount,
        type,
        sender: type === 'INCOME' ? counterparty : undefined,
        receiver: type !== 'INCOME' ? counterparty : undefined,
        balance,
        reference,
        provider: 'EVCPlus',
        category,
        confidenceScore: 0.95,
        timestamp: timestamp || Date.now(),
      };
    } catch {
      return null;
    }
  }
}
